### PR TITLE
agate: support write batch

### PR DESCRIPTION
**Pull request by GanZiheng**
Signed-off-by: GanZiheng [ganziheng98@gmail.com](mailto:ganziheng98@gmail.com)

### CODE REVIEW

File: src/batch.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        let mut new_txn = self.core.new_transaction(true, self.is_managed);
+        new_txn.commit_ts = self.commit_ts;
+        let txn = std::mem::replace(&mut self.txn, new_txn);
+        txn.commit()?;
+
+        Ok(())
+    }
+
+    fn delete(&mut self, key: Bytes) -> Result<()> {
+        if let Err(err) = self.txn.delete(key.clone()) {
```
[the remark this answers is not in the document]
Comment (GanZiheng) [reply src/batch.rs:102]: I think we should also set `self.err` here. I will update this.

File: src/batch.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+                self.err = Some(err.clone());
+                return Err(err);
+            }
+        } else {
+            return Ok(());
+        }
+
+        self.commit()?;
+
+        if let Err(err) = self.txn.set_entry(entry) {
```
Comment (Connor1996) [inline src/batch.rs:81]: I don't understand, why `set_entry` twice

Comment (GanZiheng) [reply src/batch.rs:81]: At first `set_entry`, we may get `TxnTooBig` error. In that case, we should commit and then `set_entry` again.

**Review by Connor1996 — COMMENTED** [review]
the coverage is not adequate

File: src/db.rs [hunk 3]
```diff
@@ -408,10 +409,14 @@ impl Core {
...
-            }
-            builder.add(iter.key(), &vs, 0); // TODO: support vlog length
+            let vlog_len = if vs.meta & value::VALUE_POINTER != 0 {
+                let mut vp = ValuePointer::default();
+                vp.decode(&vs.value);
+                vp.len
+            } else {
+                0
+            };
+            builder.add(iter.key(), &vs, vlog_len); // TODO: support vlog length
```
Comment (Connor1996) [inline src/db.rs:419]: remove the TODO

**Review by Connor1996 — APPROVED** [review]
rest LGTM

**Pull request merged**